p2p/client_test.go

```diff
@@ -93,25 +95,72 @@ func TestDiscovery(t *testing.T) {
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	clients := startTestNetwork(ctx, t, 5, map[int][]int{
-		1: []int{0},
-		2: []int{0},
-		3: []int{1},
-		4: []int{2},
+	clients := startTestNetwork(ctx, t, 5, map[int]hostDescr{
+		1: hostDescr{conns: []int{0}, chainID: "ORU2"},
+		2: hostDescr{conns: []int{0}, chainID: "ORU2"},
+		3: hostDescr{conns: []int{1}, chainID: "ORU1"},
+		4: hostDescr{conns: []int{2}, chainID: "ORU1"},
 	}, logger)
 
-	clients[1].chainID = "ORU2"
-	clients[2].chainID = "ORU2"
-	clients[3].chainID = "ORU1"
-	clients[4].chainID = "ORU1"
-
 	// wait for clients to finish refreshing routing tables
 	clients.WaitForDHT()
 
 	assert.Contains(clients[3].host.Network().Peers(), clients[4].host.ID())
 	assert.Contains(clients[4].host.Network().Peers(), clients[3].host.ID())
 }
 
+func TestGossiping(t *testing.T) {
+	assert := assert.New(t)
+	logger := &TestLogger{t}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// network connections topology: 3<->1<->0<->2<->4
+	clients := startTestNetwork(ctx, t, 5, map[int]hostDescr{
+		0: hostDescr{conns: []int{}, chainID: "2"},
+		1: hostDescr{conns: []int{0}, chainID: "1", realKey: true},
+		2: hostDescr{conns: []int{0}, chainID: "1", realKey: true},
+		3: hostDescr{conns: []int{1}, chainID: "2", realKey: true},
+		4: hostDescr{conns: []int{2}, chainID: "2", realKey: true},
+	}, logger)
+
+	// wait for clients to finish refreshing routing tables
+	clients.WaitForDHT()
+
+	var expectedMsg = []byte("foobar")
+	var wg sync.WaitGroup
+
+	// ensure that Tx is delivered to client
+	assertRecv := func(tx *Tx) {
+		logger.Debug("received tx", "body", string(tx.Data))
+		assert.Equal(expectedMsg, tx.Data)
+		wg.Done()
+	}
+	wg.Add(2)
+	clients[0].SetTxHandler(assertRecv)
+	clients[3].SetTxHandler(assertRecv)
+
+	// ensure that Tx is not delivered to client
+	assertNotRecv := func(*Tx) {
+		t.Fatal("unexpected Tx received")
+	}
+	clients[1].SetTxHandler(assertNotRecv)
+	clients[2].SetTxHandler(assertNotRecv)
+	clients[4].SetTxHandler(assertNotRecv)
+
+	// this sleep is required for pubsub to "propagate" subscription information
+	// TODO(tzdybal): is there a better way to wait for readiness?
+	time.Sleep(1 * time.Second)
```

Comment:
Nope) Even tests in libp2p use sleep but with much less time) Though I've seen discussions about fixing this several time thought more than a year)

Reply:
1s is empirical result for GitHub actions ;) Locally even 100ms works.